crates/sui-core/src/congestion_tracker.rs

```diff
@@ -0,0 +1,331 @@
+    pub fn get_suggested_gas_prices(&self, transaction: &TransactionData) -> Option<u64> {
+        self.get_suggested_gas_price_for_objects(
+            transaction
+                .shared_input_objects()
+                .into_iter()
+                .filter(|id| id.mutable)
+                .map(|id| id.id),
+        )
+    }
+}
+
+impl CongestionTracker {
+    fn process_per_checkpoint_events(
+        &self,
+        now: CheckpointTimestamp,
+        congestion_events: &[(u64, Vec<ObjectID>)],
+        cleared_events: &[(u64, Vec<ObjectID>)],
+    ) {
+        let congestion_info_map =
+            self.compute_per_checkpoint_congestion_info(now, congestion_events, cleared_events);
+        self.process_checkpoint_congestion(congestion_info_map);
+    }
+
+    fn get_suggested_gas_price_for_objects(
+        &self,
+        objects: impl Iterator<Item = ObjectID>,
+    ) -> Option<u64> {
+        let mut clearing_price = None;
+        for object_id in objects {
+            if let Some(info) = self.get_congestion_info(object_id) {
+                let clearing_price_for_object = match info
+                    .last_success_time
+                    .cmp(&Some(info.last_cancellation_time))
+                {
+                    std::cmp::Ordering::Greater => {
+                        // there were no cancellations in the most recent checkpoint,
+                        // so the object is probably not congested any more
+                        None
+                    }
+                    std::cmp::Ordering::Less => {
+                        // there were no successes in the most recent checkpoint. This should be a rare case,
+                        // but we know we will have to bid at least as much as the highest cancelled price.
+                        Some(info.highest_cancelled_gas_price)
+                    }
+                    std::cmp::Ordering::Equal => {
+                        // there were both successes and cancellations.
+                        info.lowest_executed_gas_price
+                    }
+                };
+                clearing_price = std::cmp::max(clearing_price, clearing_price_for_object);
+            }
+        }
+        clearing_price
+    }
+
+    fn compute_per_checkpoint_congestion_info(
+        &self,
+        now: CheckpointTimestamp,
+        congestion_events: &[(u64, Vec<ObjectID>)],
+        cleared_events: &[(u64, Vec<ObjectID>)],
+    ) -> HashMap<ObjectID, CongestionInfo> {
+        let mut congestion_info_map: HashMap<ObjectID, CongestionInfo> = HashMap::new();
+
+        for (gas_price, objects) in congestion_events {
+            for object in objects {
+                match congestion_info_map.entry(*object) {
+                    Entry::Occupied(entry) => {
+                        entry.into_mut().update_for_cancellation(now, *gas_price);
+                    }
+                    Entry::Vacant(entry) => {
+                        let info = CongestionInfo {
```

Comment:
Should this branch take `self.get_congestion_info(*object)` into consideration?

Reply:
I don't think so since that would be from the previous checkpoint. We consult it in the second loop to determine whether cleared events should be added, as explained in the comment.

Reply:
If `previous checkpoint` is not considered, meaning all the CongestionInfo are just created in this loop, `does "now" ever differ from self.last_cancellation_time` when calling update_for_cancellation (from my other comment)? Or do I miss something?

Reply:
i'm not sure I understand the question - I'm setting `last_cancellation_time` equal to `now` so no it would not differ?